File: terminator-application/src/main/java/io/quartic/terminator/WebsocketEndpoint.java
package io.quartic.terminator;

import com.codahale.metrics.annotation.ExceptionMetered;
import com.codahale.metrics.annotation.Metered;
import com.codahale.metrics.annotation.Timed;
import com.fasterxml.jackson.core.JsonProcessingException;
import io.quartic.common.server.ResourceManagingEndpoint;
import io.quartic.terminator.api.FeatureCollectionWithTerminationId;
import org.slf4j.Logger;
import rx.Observable;
import rx.Subscription;

import javax.websocket.Session;

import static io.quartic.common.serdes.ObjectMappers.OBJECT_MAPPER;
import static org.slf4j.LoggerFactory.getLogger;

@Metered
@Timed
@ExceptionMetered
public class WebsocketEndpoint extends ResourceManagingEndpoint<Subscription> {
    private static final Logger LOG = getLogger(WebsocketEndpoint.class);

    private final Observable<FeatureCollectionWithTerminationId> observable;

    public WebsocketEndpoint(Observable<FeatureCollectionWithTerminationId> observable) {
        this.observable = observable;
    }

    @Override
    protected Subscription createResourceFor(Session session) {
        return observable.subscribe((fcwdi) -> {
            try {
                session.getAsyncRemote().sendText(OBJECT_MAPPER.writeValueAsString(fcwdi));
            } catch (JsonProcessingException e) {
                LOG.error("Error producing JSON", e);
            }
        });
    }

    @Override
    protected void releaseResource(Subscription subscription) {
        subscription.unsubscribe();
    }
}

File: terminator-application/src/main/kotlin/io/quartic/terminator/WebsocketEndpoint.kt
package io.quartic.terminator

import com.codahale.metrics.annotation.ExceptionMetered
import com.codahale.metrics.annotation.Metered
import com.codahale.metrics.annotation.Timed
import com.fasterxml.jackson.core.JsonProcessingException
import io.quartic.common.logging.logger
import io.quartic.common.serdes.objectMapper
import io.quartic.common.server.ResourceManagingEndpoint
import io.quartic.terminator.api.FeatureCollectionWithTerminationId
import rx.Observable
import rx.Subscription
import javax.websocket.Session

@Metered
@Timed
@ExceptionMetered
class WebsocketEndpoint(private val observable: Observable<FeatureCollectionWithTerminationId>) : ResourceManagingEndpoint<Subscription>() {
    private val LOG by logger()

    override fun createResourceFor(session: Session) = observable.subscribe { fcwdi ->
        try {
            session.asyncRemote.sendText(objectMapper().writeValueAsString(fcwdi))
        } catch (e: JsonProcessingException) {
            LOG.error("Error producing JSON", e)
        }
    }

    override fun releaseResource(subscription: Subscription) = subscription.unsubscribe()
}
